### PR TITLE
[WIP] [Mangling] Mangle more standard substitutions

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce more standard substitutions to the mangling scheme, to better compress manglings that involve common standard library types.

Implements rdar://problem/40468532.

**Note**: This shaves ~472k off the size of the unstripped standard library binary.

### CODE REVIEW

**Comment by jckarter** [comment]
@eeckstein With the new substitution mangling scheme, is `S0` still a valid normal substitution? If not, we could squeeze eleven more standard substitutions using `S[0-9_]`.

**Comment by jrose-apple** [comment]
It seems useful to me to reserve some expansion capacity—for example, S_XY could mean "standard substitution of length 2, value XY". Even without that there's still a few letters left.

File: include/swift/Demangling/StandardTypesMangling.def [hunk 1]
```diff
@@ -14,21 +14,55 @@
...
+STANDARD_TYPE(Protocol, F, Decodable)
+STANDARD_TYPE(Protocol, B, Sequence)
+STANDARD_TYPE(Protocol, G, BidirectionalCollection)
+STANDARD_TYPE(Protocol, H, RandomAccessCollection)
+STANDARD_TYPE(Protocol, h, MutableCollection)
+STANDARD_TYPE(Protocol, J, RangeReplaceableCollection)
+STANDARD_TYPE(Protocol, j, Equatable)
+STANDARD_TYPE(Protocol, K, Comparable)
+STANDARD_TYPE(Protocol, k, Hashable)
+STANDARD_TYPE(Protocol, L, Character)
```
Comment (jrose-apple) [inline include/swift/Demangling/StandardTypesMangling.def]: This is a struct, isn't it?

**Review by jrose-apple — COMMENTED** [review]
Maybe this is silly of me but I'd prefer if the letters lined up with the protocols as much as possible too, like they do for many of the structs.

```
STANDARD_TYPE(Protocol, B, BinaryFloatingPoint)
STANDARD_TYPE(Protocol, C, Character) // not actually a protocol??
STANDARD_TYPE(Protocol, E, Encodable)
STANDARD_TYPE(Protocol, e, Decodable) // to go with Encodable; Set becomes 'h', to go with Hashable
STANDARD_TYPE(Protocol, F, FloatingPoint)
STANDARD_TYPE(Protocol, G, RandomNumberGenerator)
// 'g' is open
STANDARD_TYPE(Protocol, H, Hashable)
// 'h' for Set, to go with Hashable
// 'J' is open
STANDARD_TYPE(Protocol, j, Numeric) // a stretch, comes from i-j-k as indexes
STANDARD_TYPE(Protocol, K, BidirectionalCollection) // "kollection", plus pairing with RAC
STANDARD_TYPE(Protocol, k, RandomAccessCollection) // "kollection", plus pairing with BC
STANDARD_TYPE(Protocol, L, Comparable) // "less than"
STANDARD_TYPE(Protocol, l, Collection) // co'l'lection, plus right before MC
STANDARD_TYPE(Protocol, M, MutableCollection)
STANDARD_TYPE(Protocol, m, RangeReplaceableCollection) // pairing with MC
// 'o' is open
STANDARD_TYPE(Protocol, Q, Equatable) // e'q'uatable
STANDARD_TYPE(Protocol, T, Sequence) // pairing with IteratorProtocol
STANDARD_TYPE(Protocol, t, IteratorProtocol) // i't'erator
STANDARD_TYPE(Protocol, U, UnsignedInteger)
STANDARD_TYPE(Protocol, X, RangeExpression) // e'x'pression
STANDARD_TYPE(Protocol, x, Strideable) // pairing with RangeExpression
STANDARD_TYPE(Protocol, Y, RawRepresentable) // arbitrary
STANDARD_TYPE(Protocol, y, StringProtocol) // arbitrary
STANDARD_TYPE(Protocol, Z, BinaryInteger) // Z, from math
STANDARD_TYPE(Protocol, z, SignedInteger) // Z, from math
```

**Comment by jckarter** [comment]
Did you look at what impact this has on binary sizes on a corpus like the compatibility suite? The effect on standard library binary size may not be representative due to the fact that these types are all defined there, but the standard library's size is somewhat less important in the future for Apple platforms when it's in the shared cache.

**Comment by DougGregor** [comment]
I did not look at binary sizes across the compatibility suite. I'm going to call "perfect is the enemy of the good" on this optimization problem and focus on getting it merged. Y'all didn't comment on the (de-/re-/)mangler code changes ;)

**Comment by DougGregor** [comment]
*Sigh*, I'm taking @jrose-apple 's attempt at lining up letters with names, though. Updates coming.

**Comment by DougGregor** [comment]
Note that `o` and `C` are for Objective-C and C imports.

**Comment by jckarter** [comment]
@DougGregor Sorry, I didn't mean to raise the issue saying you should redo your work or try to adapt the set of substitutions based on some statistical model or anything like that, only that I think the size savings there would be a more interesting and compelling message than the size impact on the stdlib alone.

**Comment by jrose-apple** [comment]
Okay, okay, I'll admit I'm not sure why some of those manglings are disallowing the standard substitutions, and comments as to why would be nice.

**Comment by jrose-apple** [comment]
In that case I suggest 'g' for "grapheme cluster" for Character.

**Comment by DougGregor** [comment]
`g` is already used as a shortcut for optional types:

  type ::= type 'Sg'                         // optional type, shortcut for: type 'ySqG'

**Comment by jrose-apple** [comment]
…I guess that makes it 'J', and the alphabet is filled after all!

**Comment by eeckstein** [comment]
lgtm

**Pull request merged**